### PR TITLE
Identity Device Code , Client Certificate and VS Code Credential updates

### LINKED ISSUE #11649: [BUG] VSCodeCredential throws exception when DAC is run in Container
Get this error when using DAC in container

For users that don't need VSCodeCredential, include excludeVSCodeCredential in your DAC builder and with beta4, you must also excludeManagedIdentityCredential() #11373 and excludeSharedTokenCacheCredential() is included to ensure we use CliCredential.

```
    DefaultAzureCredential credential = new DefaultAzureCredentialBuilder()
      .excludeManagedIdentityCredential()
      .excludeSharedTokenCacheCredential()
      .excludeVSCodeCredential()
      .build();
```

```
Native library (linux-x86-64/liblibsecret-1.so) not found in resource path ([file:/usr/app/target/sdk-1.0-SNAPSHOT.jar])
sdk_1  |        at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:302)
sdk_1  |        at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:455)
sdk_1  |        at com.sun.jna.Library$Handler.<init>(Library.java:192)
sdk_1  |        at com.sun.jna.Native.load(Native.java:596)
sdk_1  |        at com.sun.jna.Native.load(Native.java:570)
sdk_1  |        at com.microsoft.aad.msal4jextensions.persistence.linux.ISecurityLibrary.<clinit>(ISecurityLibrary.java:18)
sdk_1  |        at com.azure.identity.implementation.LinuxKeyRingAccessor.read(LinuxKeyRingAccessor.java:50)
sdk_1  |        at com.azure.identity.implementation.LinuxKeyRingAccessor.read(LinuxKeyRingAccessor.java:70)
sdk_1  |        at com.azure.identity.implementation.VisualStudioCacheAccessor.getCredentials(VisualStudioCacheAccessor.java:132)
sdk_1  |        at com.azure.identity.implementation.IdentityClient.authenticateWithVsCodeCredential(IdentityClient.java:539)
sdk_1  |        at com.azure.identity.VisualStudioCodeCredential.lambda$getToken$2(VisualStudioCodeCredential.java:76)
sdk_1  |        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44)
sdk_1  |        at reactor.core.publisher.Mono.subscribe(Mono.java:4218)
sdk_1  |        at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:75)
sdk_1  |        at reactor.core.publisher.Operators.complete(Operators.java:135)
sdk_1  |        at reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:45)
sdk_1  |        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
sdk_1  |        at reactor.core.publisher.Mono.subscribe(Mono.java:4218)
sdk_1  |        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:418)
sdk_1  |        at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath(FluxIterable.java:267)
sdk_1  |        at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:225)
sdk_1  |        at reactor.core.publisher.FluxFlatMap$FlatMapMain.onSubscribe(FluxFlatMap.java:363)
sdk_1  |        at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:161)
sdk_1  |        at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:86)
sdk_1  |        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
sdk_1  |        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
sdk_1  |        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
sdk_1  |        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
sdk_1  |        at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
sdk_1  |        at reactor.core.publisher.MonoDelaySubscription.accept(MonoDelaySubscription.java:52)
sdk_1  |        at reactor.core.publisher.MonoDelaySubscription.accept(MonoDelaySubscription.java:33)
sdk_1  |        at reactor.core.publisher.FluxDelaySubscription$DelaySubscriptionOtherSubscriber.onNext(FluxDelaySubscription.java:123)
sdk_1  |        at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117)
sdk_1  |        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
sdk_1  |        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
sdk_1  |        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
sdk_1  |        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
sdk_1  |        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
sdk_1  |        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
sdk_1  |        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
sdk_1  |        at java.lang.Thread.run(Thread.java:748)
sdk_1  |        Suppressed: java.lang.UnsatisfiedLinkError: liblibsecret-1.so: cannot open shared object file: No such file or directory
sdk_1  |                at com.sun.jna.Native.open(Native Method)
sdk_1  |                at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:191)
sdk_1  |                ... 40 common frames omitted
sdk_1  |        Suppressed: java.lang.UnsatisfiedLinkError: liblibsecret-1.so: cannot open shared object file: No such file or directory
sdk_1  |                at com.sun.jna.Native.open(Native Method)
sdk_1  |                at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:204)
sdk_1  |                ... 40 common frames omitted
sdk_1  |        Suppressed: java.io.IOException: Native library (linux-x86-64/liblibsecret-1.so) not found in resource path ([file:/usr/app/target/sdk-1.0-SNAPSHOT.jar])
sdk_1  |                at com.sun.jna.Native.extractFromResourcePath(Native.java:1095)
sdk_1  |                at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:276)
sdk_1  |                ... 40 common frames omitted
```

0. az login
1. create new maven project
2. add to pom.xml

```
<?xml version="1.0" encoding="UTF-8" ?>

<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>demo</groupId>
  <artifactId>sdk</artifactId>
  <version>1.0-SNAPSHOT</version>

  <name>sdk</name>
  <!-- FIXME change it to the project's website -->
  <url>http://www.example.com</url>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
  </properties>

  <dependencies>

    <dependency>
      <groupId>io.github.cdimascio</groupId>
      <artifactId>java-dotenv</artifactId>
      <version>5.2.0</version>
    </dependency>
    
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-api</artifactId>
      <version>1.7.30</version>
    </dependency>

    <dependency>
      <groupId>ch.qos.logback</groupId>
      <artifactId>logback-classic</artifactId>
      <version>1.2.3</version>
    </dependency>

    <dependency>
      <groupId>com.azure</groupId>
      <artifactId>azure-storage-blob</artifactId>
      <version>12.6.1</version>
    </dependency>


    <dependency>
      <groupId>com.azure</groupId>
      <artifactId>azure-identity</artifactId>
      <version>1.1.0-beta.4</version>
    </dependency>

    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.11</version>
      <scope>test</scope>
    </dependency>
  </dependencies>

  <build>
    <pluginManagement>
      <!-- lock down plugins versions to avoid using Maven defaults (may be moved to parent pom) -->
      <plugins>
        <!-- clean lifecycle, see https://maven.apache.org/ref/current/maven-core/lifecycles.html#clean_Lifecycle -->
        <plugin>
          <artifactId>maven-clean-plugin</artifactId>
          <version>3.1.0</version>
        </plugin>
        <!-- default lifecycle, jar packaging: see https://maven.apache.org/ref/current/maven-core/default-bindings.html#Plugin_bindings_for_jar_packaging -->
        <plugin>
          <artifactId>maven-resources-plugin</artifactId>
          <version>3.0.2</version>
        </plugin>
        <plugin>
          <artifactId>maven-compiler-plugin</artifactId>
          <version>3.8.0</version>
        </plugin>
        <plugin>
          <artifactId>maven-surefire-plugin</artifactId>
          <version>2.22.1</version>
        </plugin>
        <plugin>
          <artifactId>maven-jar-plugin</artifactId>
          <version>3.0.2</version>
        </plugin>
        <plugin>
          <artifactId>maven-install-plugin</artifactId>
          <version>2.5.2</version>
        </plugin>
        <plugin>
          <artifactId>maven-deploy-plugin</artifactId>
          <version>2.8.2</version>
        </plugin>
        <!-- site lifecycle, see https://maven.apache.org/ref/current/maven-core/lifecycles.html#site_Lifecycle -->
        <plugin>
          <artifactId>maven-site-plugin</artifactId>
          <version>3.7.1</version>
        </plugin>
        <plugin>
          <artifactId>maven-project-info-reports-plugin</artifactId>
          <version>3.0.0</version>
        </plugin>
        <plugin>
          <artifactId>maven-assembly-plugin</artifactId>
          <executions>
            <execution>
              <phase>package</phase>
              <goals>
                <goal>single</goal>
              </goals>
            </execution>
          </executions>
          <configuration>
            <archive>
              <manifest>
                <addClasspath>true</addClasspath>
                <mainClass>demo.App</mainClass>
              </manifest>
            </archive>
            <descriptorRefs>
              <descriptorRef>jar-with-dependencies</descriptorRef>
            </descriptorRefs>
            <appendAssemblyId>false</appendAssemblyId>
          </configuration>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
</project>
```

Code
```java
package demo;

import ch.qos.logback.classic.Logger;
import com.azure.identity.DefaultAzureCredential;
import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.BlobServiceClientBuilder;
import io.github.cdimascio.dotenv.Dotenv;
import org.slf4j.LoggerFactory;

public class App {
  private static final Logger LOGGER = (Logger) LoggerFactory.getLogger("root");

  private static final Dotenv ENV = Dotenv
    .configure()
    .ignoreIfMissing()
    .systemProperties()
    .directory("../../")
    .load();

  public static void main(String[] args) {
    DefaultAzureCredential credential = new DefaultAzureCredentialBuilder()
      .excludeManagedIdentityCredential()
      .excludeSharedTokenCacheCredential()
      //.excludeVSCodeCredential()
      .build();

    BlobServiceClient blobStorageClient = new BlobServiceClientBuilder()
      .endpoint(ENV.get("AZURE_STORAGE_BLOB_URL"))
      .credential(credential)
      .buildClient();

    BlobContainerClient blobContainerClient = blobStorageClient.getBlobContainerClient(
      ENV.get("AZURE_STORAGE_CONTAINER_NAME")
    );

    Boolean exists = blobContainerClient.exists();
    LOGGER.info("Container exists: {}", exists);
  }
}

```

logback.xml
```
<?xml version="1.0" encoding="UTF-8" ?>
<configuration>
    <appender name="Console" class="ch.qos.logback.core.ConsoleAppender">
        <layout class="ch.qos.logback.classic.PatternLayout">
            <Pattern>
        %white(%d{ISO8601}) %white(%-5level) [%white(%t)] %white(%logger{100}): %msg%n%throwable
      </Pattern>
        </layout>
    </appender>
    <root level="trace">
        <appender-ref ref="Console" />
    </root>
    <logger name="reactor.netty" level="info" />
    <logger name="io.netty" level="info" />
</configuration>
```

docker-compose.xml
```yaml
version: '3.4'

services:
  sdk:
    env_file: ../../.env
    volumes: 
      - "${HOME}/.m2:/root/.m2"
      - "${HOME}/.azure:/home/developer/.azure"
    image: sdk
    build: .
```

Dockerfile
```
FROM maven:3-jdk-8-alpine AS build-env
WORKDIR /usr/app/
COPY pom.xml .
RUN mvn dependency:go-offline

COPY src/ /usr/app/src
RUN mvn package -Dmaven.test.skip=true assembly:single

# Build runtime image
FROM openkbs/jre-mvn-py3
RUN pip3 install --user azure-cli
ENV PATH /home/developer/.local/bin/:$PATH

WORKDIR /usr/app/
COPY --from=build-env /usr/app/ .
ENTRYPOINT ["java", "-jar", "target/sdk-1.0-SNAPSHOT.jar"]
```

azsdke2e

### LINKED ISSUE #16068: [FEATURE REQ] Apply Device Code Credential Consistency Review Feedback
- [ ] Make Client Id optional. Default to Developer Single Sign On Id (Azure CLI Id for now.) 
- [ ] Email arch board regarding Tenant Id validation.

### LINKED ISSUE #16936: Address Client Certificate Credential Consistency Review feedback
Following changes need to be done, based on the review discussion:

- [ ] Update Tenant ID Validation Message.
- [ ] Rename includeX5c param to send certificate chain.
- [ ]  Check credential behavior when scope is passed as null (How does MSAL handle it ?).
- [ ] Follow up on Validations for Cert path and client ID (remove them, if they're false positives)
- [ ] Check possibility of adding ClientAuthenticationException in CCC. (Avoid breaking changes)

### PR DESCRIPTION
**Pull request by g2vinay**
Fixes #11649
Resolves the VS code Docker Container Env Issue.

Fixes #16068 
Applies the Device Code Credential consistency review feedback.

Fixes #16936
Applies the Client Certificate Credential consistency review feedback.

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityConstants.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.identity.implementation.util;
+
+public class IdentityConstants {
+    /**
+     * The Single Sign On ID.
+     */
+    public static final String DEVELOPER_SINGLE_SIGN_ON_ID = "04b07795-8ddb-461a-bbee-02f9e1bf7b46";
```
Comment (jianghaolu) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityConstants.java:10]: Single sign on ID? This is the client ID of Azure CLI.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java [hunk 2]
```diff
@@ -131,7 +131,7 @@ public String getCredentials(String serviceName, String accountName) {
 
                 byte[] readCreds = keyRingAccessor.read();
                 credential = new String(readCreds, StandardCharsets.UTF_8);
-            } catch (Exception e) {
+            } catch (Exception | UnsatisfiedLinkError e) {
```
Comment (jianghaolu) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java:134]: Why not use `Throwable`?

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java:134]: Throwable will work too, but didn't want to aggressively block other Error types which may qualify to fail the credential.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityConstants.java (hunk 1 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/IdentityConstants.java:10]: Yeah the CLI ID is being used temporarily here in other languages too, until we get other Client Id.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ClientCertificateCredentialBuilder.java [hunk 3]
```diff
@@ -99,11 +97,12 @@ ClientCertificateCredentialBuilder enablePersistentCache() {
...
      *
-     * @param includeX5c the flag to indicate if x5c should be sent as part of authentication request.
+     * @param sendCertificateChain the flag to indicate if certificate chain should be sent as part of authentication
+     * request.
      * @return An updated instance of this builder.
      */
-    public ClientCertificateCredentialBuilder includeX5c(boolean includeX5c) {
-        this.identityClientOptions.setIncludeX5c(includeX5c);
+    public ClientCertificateCredentialBuilder sendCertificateChain(boolean sendCertificateChain) {
+        this.identityClientOptions.setIncludeX5c(sendCertificateChain);
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/ClientCertificateCredentialBuilder.java:105]: For internal consistency you should consider changing the name on the identityClientOptions too

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java [hunk 4]
```diff
@@ -131,7 +131,7 @@ public String getCredentials(String serviceName, String accountName) {
...
                 throw logger.logExceptionAsError(new CredentialUnavailableException(
                         "Failed to read Vs Code credentials from Linux Key Ring.", e));
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/VisualStudioCacheAccessor.java:136]: I would expect all strings going into logs would refer to it as "VS Code"?

**Review by JonathanGiles — APPROVED** [review]
Minor nits only

**Review by schaabs — APPROVED** [review]

**Pull request merged**